### ISSUE TITLE
[ILM] The downsample action runs before the allocate and migrate actions

### ISSUE THREAD

**Issue by andreidan** [elastic/elasticsearch#96734]
### Elasticsearch Version

8.7+

### Java Version

_bundled_

### OS Version

Darwin

### Problem Description

In the warm and cold phases the downsample action runs before allocate and migrate https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java#L69

This means that if I define a policy with hot and warm like so:
```
      "phases" : {
        "hot" : {
          "actions" : {
            "rollover" : {
              "max_primary_shard_size" : "50gb",
              "max_age" : "30d",
              "max_docs" : 2000000000
            }
          },
        "warm" : {
          "min_age" : "2d",
          "actions" : {
            "allocate" : {
              "number_of_replicas" : 0,
              "include" : { },
              "exclude" : { },
              "require" : { }
            },
            "downsample" : {
              "fixed_interval" : "10h"
            }
          }
        }
```

And index would start in hot, wait to meet the rollover condition, rollover, and then **while still in the hot tier** it will
run the downsample action. Note that **the source index** will still be in the hot tier whilst downsample runs. (the downsample index will also be in hot for reasons detailed here: https://github.com/elastic/elasticsearch/issues/96733)

After downsample is completed, the downsampled index will migrate to the warm tier.

Note that this essentially makes the downsample action definition in the ILM hot and warm phase identical from an execution perspective (there's no difference as to which tier downsample will run on).

Similarly in a deployment with hot/warm/cold tiers, if downsample is defined in the `cold` phase the operation will run on the **warm tier** (i.e. source index is in warm, runs downsample, when downsample is completed it migrates to cold - as it's defined to run before the `migrate` action in cold https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java#L79 )

This might be desired in which case we should document this behaviour as it's a bit surprising. But maybe downsample should run after the migrate action? 

### Steps to Reproduce

Run downsample with ILM and check the logs for the source index i.e notice the `downsample-*` index is created whilst the source index is in the tier prior to the tier where the downsample action is defined 
https://www.elastic.co/guide/en/elasticsearch/reference/current/downsampling-ilm.html

**Comment by andreidan** [comment]
I've tagged ILM and TSDB on this issue (as ILM is where possible changes would need to be made) however it'd be up to the  @elastic/es-analytics-geo team to define the desired behaviour here.

**Comment by leighatelastic** [comment]
Can we have an update on this item please?

**Comment by martijnvg** [comment]
@andreidan I think it makes sense that the `downsample` action runs after `allocate` and `migrate` actions. But I recall this requires a breaking change in ILM? Additionally is this more a storage engine task or a data management task?

**Comment by andreidan** [comment]
@martijnvg yeah, we have a PR out there that we need to pick up and drive to completion https://github.com/elastic/elasticsearch/pull/97745

**Comment by andreidan** [comment]
Thanks for looking into this Mary. 

I think reordering might still makes sense due to the following additional cons for the current approach: 

1. the definition of the downsample action in hot and warm is identical from an execution perspective which is confusing IMO (i.e. in both cases the downsampling executes in the hot tier)
2. Downsampling waits for the time series [_end_ timestamp to lapse](https://www.elastic.co/docs/manage-data/data-store/data-streams/time-series-data-stream-tsds#time-bound-indices) before executing. A definition of `hot: rollover, warm: downsample`  will _currently_ keep the source index in the `hot` tier for potentially considerable amount of time before downsampling it and transitioning it to the `warm` tier (up to 30 minutes by default)

I'll defer to @martijnvg for additional insights from users which I agree should guide progress here. We could also choose to only document this behaviour..

**Comment by gmarouli** [comment]
Thanks for the input @andreidan!

**Comment by martijnvg** [comment]
Thanks @gmarouli and @andreidan for your analysis here. Given that we haven't seen feedback from users around this issue, I suggest that for now we only document this behavior? 

Additionally to what @gmarouli mentioned, having more performant hardware for downsampling is an advantage, given that downsampling is very resource intensive (both CPU and IO). I have seen cases where users downsample to 1 minute intervals aggressively, which would become much slower if that were to happen in a hot tier. This downsample configuration is more resource intensive compared to downsampling to 1 hour, given that target index will still hold a lot of documents, and so relatively most resource utilization will be spent on indexing.

**Comment by gmarouli** [comment]
Thank for the input @martijnvg , I agree. Let's close it for now and if we get different feedback from users we can reopen it.
